detox/src/devices/android/AAPT.js

```diff
@@ -1,23 +1,44 @@
 const _ = require('lodash');
 const Environment = require('../../utils/environment');
+const fs = require('fs-extra');
 const path = require('path');
+const which = require('which');
 const exec = require('../../utils/exec').execWithRetriesAndLogs;
 const escape = require('../../utils/pipeCommands').escape.inQuotedString;
 const egrep = require('../../utils/pipeCommands').search.fragment;
 const fsext = require('../../utils/fsext');
-
 class AAPT {
-
   constructor() {
     this.aaptBin = null;
   }
 
   async _prepare() {
-    if (!this.aaptBin) {
-      const sdkPath = Environment.getAndroidSDKPath();
-      const buildToolsDirs = await fsext.getDirectories(path.join(sdkPath, 'build-tools'));
-      const latestBuildToolsVersion = _.last(buildToolsDirs);
-      this.aaptBin = path.join(sdkPath, 'build-tools', latestBuildToolsVersion, 'aapt');
+    if (this.aaptBin) {
+      return;
+    }
+
+    const sdkPath = Environment.getAndroidSDKPath();
+
+    let latestBuildToolsVersion = '';
+
+    const buildToolsDir = path.join(sdkPath, 'build-tools');
+    if (fs.pathExistsSync(buildToolsDir)) {
+      const buildToolsDirs = await fsext.getDirectories(buildToolsDir);
+      latestBuildToolsVersion = _.last(buildToolsDirs);
+    }
+
+    const buildToolsDirLatestVersion = path.join(
+      sdkPath,
+      'build-tools',
+      latestBuildToolsVersion,
+    );
+
+    this.aaptBin =
```

Comment:
One more issue emerged.
I see the same repetitive pattern across those multiple "Android tools"-related wrapper classes.
You get some path, you perform some `path.join`s, you perform some `exist`ence checks, you call `which.sync` multiple times, rinse and repeat.

I won't be repeating this message for all those three files you have changed, just examine them and get the idea.

I suggest extracting this logic to a separate class (e.g. `AndroidSDK`), thereby eliminating a single concern spread across multiple classes. You might then export an instance of it in `environment.js`.

On the low-level scale, your flow now looks better than previously.